### PR TITLE
Ability to configure WebSocket connection

### PR DESCRIPTION
**Pull request by abodalevsky**
It's proposed to add options parameter to configure WebSocket. 
```javascript
const options = {
  headers: {
    origin: 'www.example.com'
  },
  proxy: {
    useDefault: true,
    uri: 'proxy.address.com',
    user: 'name',
    password: 'p@ssw0rd'
 }
}

const ws = new WebSocket('ws://echo.websocket.org', null, options)
```

**headers** section brings support _Origin_ header it addresses issue #253. Structure is aligned with implementation for Android.

**proxy** section extends support of proxy for UWP and brings support of proxy servers for WPF.
 - _useDefault_ - is relevant for WPF and indicates that connection should be established via proxy if it was auto-discovered. It is not relevant for UWP just because it's already behavior by default.
 - _uri_, _user_, _password_ - set credentials that use to authenticate to the proxy server through HTTP header authentication.

Any of this parameters can be omitted.

### CODE REVIEW

**Comment by rozele** [comment]
@abodalevsky - can you link to the proxy implementation on Android? I don't see anything in https://github.com/facebook/react-native/blob/master/Libraries/WebSocket/WebSocket.js.

**Comment by rozele** [comment]
Or in https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java for that matter.

**Comment by abodalevsky** [comment]
I wasn't able to find it either. As I posted to discord, it uses OkHttpClient library capability, e.g. when proxy is detected WebSocket connection is tried to be established via proxy.

File: ReactWindows/ReactNative.Shared.Tests/Modules/WebSocket/WebSocketOptionsTests.cs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        [Test]
+        [Category("WebSocket")]
+        public void WebSocketOptions_NoProxy()
+        {
+            var json = @"{
+    origin: 'www.example.com'
+}";
+            var jOptions = JObject.Parse(json);
+            var options = new WebSocketOptions(jOptions);
+            Assert.IsFalse(options.UseDefaultProxy);
```
Comment (matthargett) [inline ReactWindows/ReactNative.Shared.Tests/Modules/WebSocket/WebSocketOptionsTests.cs]: Assert.That(options.UseDefaultProxy, Is.False);
// use the fluent style assertion syntax consistently throughout these new additions, please :)

**Comment by abodalevsky** [comment]
@matthargett , thanks for your review. I've addressed feedback. In addition I have updated _flow type annotation_ for `options` parameter of WebSocket.

File: ReactWindows/ReactNative.Net46/Modules/WebSocket/ProxyHelper.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            // we get wss:// (or ws://) scheme that is not all proxies handle correctly. For better autodiscovering
+            // replace wss:// to https:// (ws:// to http:// accordingly)
+
+            var normalizedAddress = new UriBuilder(address)
+            {
+                Scheme = address.Scheme == "wss" ? Uri.UriSchemeHttps : Uri.UriSchemeHttp
+            };
+
+            var proxy = WebRequest.DefaultWebProxy;
+            proxy.Credentials = CredentialCache.DefaultCredentials;
```
Comment (matthargett) [inline ReactWindows/ReactNative.Net46/Modules/WebSocket/ProxyHelper.cs]: pass this value into the ctor as ICredentials:
https://msdn.microsoft.com/en-us/library/system.net.webproxy.credentials(v=vs.110).aspx

File: ReactWindows/ReactNative.Net46/Modules/WebSocket/ProxyHelper.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            // we get wss:// (or ws://) scheme that is not all proxies handle correctly. For better autodiscovering
+            // replace wss:// to https:// (ws:// to http:// accordingly)
+
+            var normalizedAddress = new UriBuilder(address)
+            {
+                Scheme = address.Scheme == "wss" ? Uri.UriSchemeHttps : Uri.UriSchemeHttp
+            };
+
+            var proxy = WebRequest.DefaultWebProxy;
```
Comment (matthargett) [inline ReactWindows/ReactNative.Net46/Modules/WebSocket/ProxyHelper.cs]: pass the webProxy into the ctor as IWebProxy so this class can be easily unit tested with mocks.

File: ReactWindows/ReactNative.Net46/Modules/WebSocket/ProxyHelper.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            {
+                ProxyAddress = proxy.GetProxy(normalizedAddress.Uri).AbsoluteUri;
+                var defCreds = CredentialCache.DefaultCredentials as NetworkCredential;
+
+                if (defCreds == null)
+                {
+                    return;
+                }
+
+                UserName = string.IsNullOrEmpty(defCreds.Domain) ? defCreds.UserName : $"{defCreds.Domain}\\{defCreds.UserName}";
```
Comment (matthargett) [inline ReactWindows/ReactNative.Net46/Modules/WebSocket/ProxyHelper.cs]: use null coalescing operator.

File: ReactWindows/ReactNative.Net46/Modules/WebSocket/ProxyHelper.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                Password = defCreds.Password;
+            }
+        }
+
+        public ProxyHelper(IWebSocketOptions options)
+        {
+            if (options == null)
+            {
+                throw new ArgumentNullException("options", "Parameter cannot be null");
```
Comment (matthargett) [inline ReactWindows/ReactNative.Net46/Modules/WebSocket/ProxyHelper.cs]: use nameof(options) and null coalescing operator here

File: ReactWindows/ReactNative.Net46/Modules/WebSocket/WebSocketModule.cs [hunk 6]
```diff
@@ -28,13 +28,7 @@ public WebSocketModule(ReactContext reactContext)
...
         #region NativeModuleBase Overrides
 
-        public override string Name
-        {
-            get
-            {
-                return "WebSocketModule";
-            }
-        }
+        public override string Name => "WebSocketModule";
```
Comment (matthargett) [inline ReactWindows/ReactNative.Net46/Modules/WebSocket/WebSocketModule.cs]: 👍

File: ReactWindows/ReactNative.Net46/Modules/WebSocket/ProxyHelper.cs (hunk 4 above)
Comment (abodalevsky) [reply ReactWindows/ReactNative.Net46/Modules/WebSocket/ProxyHelper.cs]: there is check for `string.Empty` not for _null_

**Pull request closed without merging**